gateway-provider-security-jwt/src/main/java/org/apache/knox/gateway/provider/federation/jwt/filter/JWTFederationFilter.java

```diff
@@ -279,7 +279,14 @@ public Pair<TokenType, String> getWireToken(final ServletRequest request) throws
               // what follows the bearer designator should be the JWT token being used
               // to request or as an access token
               token = header.substring(BEARER.length());
-              parsed = Pair.of(TokenType.JWT, token);
+
+              // if this appears to be a JWT token then attempt to use it as such
+              // otherwise assume it is a passcode token
+              if (isJWT(token)) {
```

Comment:
I'm not sure I like this approach. What if `getWireToken` actually returns a `WireTokenResult` (abstract) with 2 sub-classes:
- one holds a JWTToken (tokenType = JWT)
- another one holds a Passcode token (tokenType = Passcode)

Then `getWireToken` could actually try and parse the header value and could handle `ParseException` if it's not a JWT. Please note this is still a cheap String operation in the background (i.e. no signature verification yet).

What do you think?

Reply:
hmmm, I don't really follow. getWireToken already returns Pair<TokenType, String> which is essentially the same thing, no? Are you suggesting that we refactor that to push it through the JWTToken ctor and catch an exception on every passcode use? Throwing exceptions for flow control is generally a poor practice for performance and design reasons. The current implementation only throws exceptions in exceptional conditions. Are you suggesting something other than my assumptions here?

Reply:
Please ignore my comment above, I realized it would require way more work than I anticipated in the first place and overcomplicate this simple check. Your approach  should work fine.

Reply:
Thanks for your review, as always, @smolnar82. Much appreciated!